core/src/main/java/org/bitcoinj/crypto/AnyDeterministicHierarchy.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2013 Matija Mazi.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bitcoinj.crypto;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Maps;
+import org.bitcoinj.crypto.bls.BLSDeterministicKey;
+import org.bitcoinj.crypto.bls.BLSHDKeyDerivation;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+// TODO: This whole API feels a bit object heavy. Do we really need ChildNumber and so many maps, etc?
+// TODO: Should we be representing this using an actual tree arrangement in memory instead of a bunch of hashmaps?
+
+/**
+ * <p>A DeterministicHierarchy calculates and keeps a whole tree (hierarchy) of keys originating from a single
+ * root key. This implements part of the BIP 32 specification. A deterministic key tree is useful because
+ * Bitcoin's privacy system require new keys to be created for each transaction, but managing all these
+ * keys quickly becomes unwieldy. In particular it becomes hard to back up and distribute them. By having
+ * a way to derive random-looking but deterministic keys we can make wallet backup simpler and gain the
+ * ability to hand out {@link BLSDeterministicKey}s to other people who can then create new addresses
+ * on the fly, without having to contact us.</p>
+ *
+ * <p>The hierarchy is started from a single root key, and a location in the tree is given by a path which
+ * is a list of {@link ChildNumber}s.</p>
+ */
+public class AnyDeterministicHierarchy {
+    private final Map<ImmutableList<ChildNumber>, IDeterministicKey> keys = Maps.newHashMap();
+    private final ImmutableList<ChildNumber> rootPath;
+    // Keep track of how many child keys each node has. This is kind of weak.
+    private final Map<ImmutableList<ChildNumber>, ChildNumber> lastChildNumbers = Maps.newHashMap();
```

Comment:
This class is a copy of `DeterministicKeyHierarchy` but works with the `IDeterministicKey` keys (which are any type of key) rather than one specific ECKey.

A future PR will be adding another type of key (EDDSA / ED25519) and after that more refactoring can be considered to reuse more code rather than have copies.

For now the goal is to disturb the classes that are related to payments as little as possible while adding support for keep track of other key types.